File: backend/app/api/v1/characters.py
"""
Character endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, List, Dict, Any
from pydantic import BaseModel
import logging

from app.core.database import get_db
from app.models.character import Character
from app.models.user import User
from app.models.market import MarketOrder
from app.tasks.character_sync import (
    sync_character_assets,
    sync_character_market_orders,
    sync_character_details,
)

logger = logging.getLogger(__name__)

router = APIRouter()


class CharacterResponse(BaseModel):
    """Character response model"""
    id: int
    character_id: int
    character_name: str
    corporation_id: Optional[int]
    corporation_name: Optional[str]
    alliance_id: Optional[int]
    alliance_name: Optional[str]
    security_status: Optional[str]
    birthday: Optional[str]
    gender: Optional[str]
    race_id: Optional[int]
    bloodline_id: Optional[int]
    ancestry_id: Optional[int]
    last_synced_at: Optional[str]
    
    class Config:
        from_attributes = True


@router.get("/", response_model=List[CharacterResponse])
async def list_characters(
    user_id: Optional[int] = Query(None, description="Filter by user ID"),
    corporation_id: Optional[int] = Query(None, description="Filter by corporation ID"),
    alliance_id: Optional[int] = Query(None, description="Filter by alliance ID"),
    db: Session = Depends(get_db),
):
    """
    List characters with optional filters
    
    Args:
        user_id: Filter by user ID
        corporation_id: Filter by corporation ID
        alliance_id: Filter by alliance ID
        db: Database session
    """
    try:
        query = db.query(Character)
        
        if user_id:
            query = query.filter(Character.user_id == user_id)
        
        if corporation_id:
            query = query.filter(Character.corporation_id == corporation_id)
        
        if alliance_id:
            query = query.filter(Character.alliance_id == alliance_id)
        
        characters = query.all()
        
        return characters
        
    except Exception as e:
        logger.error(f"Error listing characters: {e}")
        raise HTTPException(status_code=500, detail="Failed to list characters")


@router.get("/{character_id}", response_model=CharacterResponse)
async def get_character(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Get character details
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        return character
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting character {character_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get character")


@router.get("/{character_id}/assets")
async def get_character_assets(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Get character assets
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        # Get assets from character_data
        assets = []
        if character.character_data and "assets" in character.character_data:
            assets = character.character_data["assets"]
        
        return {
            "character_id": character_id,
            "assets": assets,
            "count": len(assets),
            "synced_at": character.character_data.get("assets_synced_at") if character.character_data else None,
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting character assets for {character_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get character assets")


@router.get("/{character_id}/market-orders")
async def get_character_market_orders(
    character_id: int,
    is_buy_order: Optional[bool] = Query(None, description="Filter by order type (true=buy, false=sell)"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    db: Session = Depends(get_db),
):
    """
    Get character's personal market orders
    
    Args:
        character_id: EVE character ID
        is_buy_order: Filter by order type
        is_active: Filter by active status
        skip: Number of records to skip
        limit: Maximum number of records to return
        db: Database session
    """
    try:
        query = db.query(MarketOrder).filter(
            MarketOrder.character_id == character_id
        )
        
        if is_buy_order is not None:
            query = query.filter(MarketOrder.is_buy_order == is_buy_order)
        
        if is_active is not None:
            query = query.filter(MarketOrder.is_active == is_active)
        
        orders = query.order_by(MarketOrder.issued.desc()).offset(skip).limit(limit).all()
        total = query.count()
        
        # Convert to response format
        order_responses = []
        for order in orders:
            order_dict = {
                "id": order.id,
                "order_id": order.order_id,
                "character_id": order.character_id,
                "type_id": order.type_id,
                "type_name": order.type_name,
                "type_icon_url": f"https://images.evetech.net/types/{order.type_id}/icon" if order.type_id else None,
                "is_buy_order": order.is_buy_order,
                "location_id": order.location_id,
                "location_name": order.location_name,
                "region_id": order.region_id,
                "region_name": order.region_name,
                "system_id": order.system_id,
                "system_name": order.system_name,
                "price": order.price,
                "volume_total": order.volume_total,
                "volume_remain": order.volume_remain,
                "min_volume": order.min_volume,
                "duration": order.duration,
                "issued": order.issued.isoformat() if order.issued else None,
                "expires": order.expires.isoformat() if order.expires else None,
                "is_active": order.is_active,
                "range_type": order.range_type,
                "range_value": order.range_value,
            }
            order_responses.append(order_dict)
        
        return {
            "items": order_responses,
            "total": total,
            "skip": skip,
            "limit": limit,
        }
        
    except Exception as e:
        logger.error(f"Error getting character market orders for {character_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get character market orders")


@router.get("/{character_id}/details")
async def get_character_details(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Get detailed character information
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        # Get details from character_data
        details = character.character_data or {}
        
        return {
            "character_id": character_id,
            "character_name": character.character_name,
            "corporation_id": character.corporation_id,
            "corporation_name": character.corporation_name,
            "alliance_id": character.alliance_id,
            "alliance_name": character.alliance_name,
            "security_status": character.security_status,
            "details": details,
            "synced_at": details.get("details_synced_at") if details else None,
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting character details for {character_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get character details")


@router.post("/{character_id}/sync/assets")
async def trigger_sync_assets(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Trigger sync of character assets
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        task = sync_character_assets.delay(character_id)
        return {
            "message": "Character assets sync queued",
            "task_id": task.id,
            "character_id": character_id,
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error queuing character assets sync: {e}")
        raise HTTPException(status_code=500, detail="Failed to queue character assets sync")


@router.post("/{character_id}/sync/market-orders")
async def trigger_sync_market_orders(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Trigger sync of character market orders
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        task = sync_character_market_orders.delay(character_id)
        return {
            "message": "Character market orders sync queued",
            "task_id": task.id,
            "character_id": character_id,
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error queuing character market orders sync: {e}")
        raise HTTPException(status_code=500, detail="Failed to queue character market orders sync")


@router.post("/{character_id}/sync/details")
async def trigger_sync_details(
    character_id: int,
    db: Session = Depends(get_db),
):
    """
    Trigger sync of character details
    
    Args:
        character_id: EVE character ID
        db: Database session
    """
    try:
        character = db.query(Character).filter(
            Character.character_id == character_id
        ).first()
        
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")
        
        task = sync_character_details.delay(character_id)
        return {
            "message": "Character details sync queued",
            "task_id": task.id,
            "character_id": character_id,
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error queuing character details sync: {e}")
        raise HTTPException(status_code=500, detail="Failed to queue character details sync")

File: frontend/src/services/characters.ts
import api from './api'

export interface Character {
  id: number
  character_id: number
  character_name: string
  corporation_id: number | null
  corporation_name: string | null
  alliance_id: number | null
  alliance_name: string | null
  security_status: number | null
  birthday: string | null
  gender: string | null
  race_id: number | null
  bloodline_id: number | null
  ancestry_id: number | null
  last_synced_at: string | null
}

export interface CharacterAsset {
  item_id: number
  type_id: number
  type_name?: string
  type_icon_url?: string
  location_id: number
  location_name?: string
  location_type?: string
  system_id?: number
  system_name?: string
  region_name?: string
  quantity: number
  flag?: number
  singleton?: boolean
  items?: CharacterAsset[] // Nested items (containers)
}

export interface CharacterMarketOrder {
  id: number
  order_id: number
  character_id: number
  type_id: number
  type_name: string | null
  type_icon_url: string | null
  is_buy_order: boolean
  location_id: number
  location_name: string | null
  region_id: number | null
  region_name: string | null
  system_id: number | null
  system_name: string | null
  price: number
  volume_total: number
  volume_remain: number
  min_volume: number | null
  duration: number | null
  issued: string | null
  expires: string | null
  is_active: boolean
  range_type: string | null
  range_value: number | null
}

export interface CharacterDetails {
  character_id: number
  character_name: string
  corporation_id: number | null
  corporation_name: string | null
  alliance_id: number | null
  alliance_name: string | null
  security_status: string | null
  details: {
    character_info?: any
    corporation_info?: any
    alliance_info?: any
    contacts?: any[]
    standings?: any[]
    wallet_balance?: number
    location?: any
    ship?: any
    skill_queue?: any[]
    skills?: any
    [key: string]: any
  }
  synced_at: string | null
}

export const charactersService = {
  /**
   * List all characters
   */
  list: async (params?: {
    user_id?: number
    corporation_id?: number
    alliance_id?: number
  }): Promise<Character[]> => {
    const queryParams = new URLSearchParams()
    if (params?.user_id) queryParams.append('user_id', params.user_id.toString())
    if (params?.corporation_id) queryParams.append('corporation_id', params.corporation_id.toString())
    if (params?.alliance_id) queryParams.append('alliance_id', params.alliance_id.toString())
    
    const response = await api.get(`/characters?${queryParams.toString()}`)
    return response.data
  },

  /**
   * Get character by ID
   */
  get: async (characterId: number): Promise<Character> => {
    const response = await api.get(`/characters/${characterId}`)
    return response.data
  },

  /**
   * Get character assets
   */
  getAssets: async (characterId: number): Promise<{
    character_id: number
    assets: CharacterAsset[]
    count: number
    synced_at: string | null
  }> => {
    const response = await api.get(`/characters/${characterId}/assets`)
    return response.data
  },

  /**
   * Get character market orders
   */
  getMarketOrders: async (
    characterId: number,
    params?: {
      is_buy_order?: boolean
      is_active?: boolean
      skip?: number
      limit?: number
    }
  ): Promise<{
    items: CharacterMarketOrder[]
    total: number
    skip: number
    limit: number
  }> => {
    const queryParams = new URLSearchParams()
    if (params?.is_buy_order !== undefined) queryParams.append('is_buy_order', params.is_buy_order.toString())
    if (params?.is_active !== undefined) queryParams.append('is_active', params.is_active.toString())
    if (params?.skip) queryParams.append('skip', params.skip.toString())
    if (params?.limit) queryParams.append('limit', params.limit.toString())
    
    const response = await api.get(`/characters/${characterId}/market-orders?${queryParams.toString()}`)
    return response.data
  },

  /**
   * Get character details
   */
  getDetails: async (characterId: number): Promise<CharacterDetails> => {
    const response = await api.get(`/characters/${characterId}/details`)
    return response.data
  },

  /**
   * Trigger sync of character assets
   */
  syncAssets: async (characterId: number): Promise<{
    message: string
    task_id: string
    character_id: number
  }> => {
    const response = await api.post(`/characters/${characterId}/sync/assets`)
    return response.data
  },

  /**
   * Trigger sync of character market orders
   */
  syncMarketOrders: async (characterId: number): Promise<{
    message: string
    task_id: string
    character_id: number
  }> => {
    const response = await api.post(`/characters/${characterId}/sync/market-orders`)
    return response.data
  },

  /**
   * Trigger sync of character details
   */
  syncDetails: async (characterId: number): Promise<{
    message: string
    task_id: string
    character_id: number
  }> => {
    const response = await api.post(`/characters/${characterId}/sync/details`)
    return response.data
  },
}

